File: src/application/songs/usecase.rs
use super::dynamodb;
use super::entity;
use crate::application::concerns::google_verification;
use crate::application::users;
use snafu::Snafu;
use std::str::FromStr;

#[derive(Debug, Snafu)]
pub enum Error {
    #[snafu(display("Failed Google token verification: {}", source))]
    GoogleVerificationError { source: google_signin::Error },

    #[snafu(display("Song cannot be created if it already has an ID"))]
    ExistingSongError,

    #[snafu(display(
        "The song ID to be modified is not equal to ID inside the song data: {} and {}",
        song_id_1,
        song_id_2
    ))]
    WrongIDError {
        song_id_1: String,
        song_id_2: String,
    },

    #[snafu(display(
    "This song was previously saved at a more recent time - overwriting is will cause data loss.",
    ))]
    OverwriteError,

    #[snafu(display("Song owner must be equal to the user persisting the song"))]
    WrongOwnerError,

    #[snafu(display("Data store failed: {}", source))]
    DatastoreError { source: dynamodb::Error },

    #[snafu(display("Song ID not found: {}", id))]
    NotFoundError { id: String },
}

#[derive(Clone)]
pub struct Usecase {
    google_verification: google_verification::GoogleVerification,
    datastore: dynamodb::DynamoDB,
}

impl Usecase {
    pub fn new(
        google_verification: google_verification::GoogleVerification,
        songs_datastore: dynamodb::DynamoDB,
    ) -> Usecase {
        Usecase {
            google_verification: google_verification,
            datastore: songs_datastore,
        }
    }

    pub async fn get_song(&self, id: &str) -> Result<entity::Song, Error> {
        validate_song_id(id)?;

        let get_song_result = self.datastore.get_song(id).await;

        match get_song_result {
            Ok(song) => Ok(song),
            Err(err) => Err(map_datastore_error(err, id)),
        }
    }

    pub async fn create_song(
        &self,
        user_id_token: &str,
        mut song: entity::Song,
    ) -> Result<entity::Song, Error> {
        let user = self.verify_user(user_id_token)?;

        if !song.is_new() {
            return Err(Error::ExistingSongError);
        }

        if song.summary.owner != user.id {
            return Err(Error::WrongOwnerError);
        }

        song.create_id();
        song.set_saved_time();

        match self.datastore.create_song(&song).await {
            Ok(()) => Ok(song),
            Err(err) => Err(map_datastore_error(err, &song.summary.id)),
        }
    }

    pub async fn update_song(
        &self,
        user_id_token: &str,
        song_id: &str,
        mut song: entity::Song,
    ) -> Result<entity::Song, Error> {
        let user = self.verify_user(user_id_token)?;

        if song_id.is_empty() {
            return Err(Error::NotFoundError { id: "".to_string() });
        }

        if song_id != song.summary.id {
            return Err(Error::WrongIDError {
                song_id_1: song_id.to_string(),
                song_id_2: song.summary.id.to_string(),
            });
        }

        if song.summary.owner != user.id {
            return Err(Error::WrongOwnerError);
        }

        self.protect_overwrite_song(&song).await?;

        song.set_saved_time();

        match self.datastore.update_song(&song).await {
            Ok(()) => Ok(song),
            Err(err) => Err(map_datastore_error(err, &song.summary.id)),
        }
    }

    fn verify_user(&self, user_id_token: &str) -> Result<users::entity::User, Error> {
        match self.google_verification.verify(user_id_token) {
            Ok(user) => Ok(user),
            Err(err) => Err(Error::GoogleVerificationError { source: err }),
        }
    }

    async fn protect_overwrite_song(&self, song_to_update: &entity::Song) -> Result<(), Error> {
        let next_last_saved_at = song_to_update
            .summary
            .last_saved_at
            .ok_or(Error::OverwriteError)?;

        let curr_song = self.get_song(&song_to_update.summary.id).await?;
        match curr_song.summary.last_saved_at {
            Some(curr_last_saved_at) => {
                // prevent overwriting a more recent save if the last saved at timestamp is greater than the current one
                // example:
                // A --> B
                //   \----->C
                //
                // Suppose I open the song at time A on two computers, make edits on both somewhat absentmindedly
                // I first save at time B - the payload for the song for the last saved at would be A
                // presume this succeeds, the server copy is now from time B and its last saved at time is also B
                //
                // Now I save another copy at time C, but the predecessor of my copy at time C was from A
                // If I just go with last write wins, then all changes from the copy at time B would be overwritten
                // So by comparing the timestamp at the save at C (A vs B), the save will fail to protect overwriting data.
                //
                // The user can then refetch the copy at time B and copy over their changes from time C (manual merge)
                // and form a copy that will be accepted by the server:
                //
                // A --> B----->B+C
                //   \----->C---/
                if curr_last_saved_at.gt(&next_last_saved_at) {
                    log::info!("Overwriting error encountered");
                    log::info!("Current saved at: {}", curr_last_saved_at.to_rfc3339());
                    log::info!("Next saved at: {}", next_last_saved_at.to_rfc3339());

                    return Err(Error::OverwriteError);
                }
            }
            None => {} // really unexpected....but don't block on this I guess
        }

        Ok(())
    }
}

fn validate_song_id(id: &str) -> Result<(), Error> {
    match uuid::Uuid::from_str(id) {
        Ok(_) => Ok(()),
        // if it's not a UUID we won't find it in the datastore
        // just short circuit and don't hit the DB
        Err(_) => Err(Error::NotFoundError { id: id.to_string() }),
    }
}

fn map_datastore_error(err: dynamodb::Error, song_id: &str) -> Error {
    match err {
        dynamodb::Error::NotFoundError => Error::NotFoundError {
            id: song_id.to_string(),
        },
        dynamodb::Error::GenericDynamoError { .. }
        | dynamodb::Error::MalformedDataError { .. }
        | dynamodb::Error::SongSerializationError { .. } => Error::DatastoreError { source: err },
    }
}

File: go-rewrite/src/song/usecase/usecase.go
package songusecase

import (
	"context"
	"github.com/cockroachdb/errors/markers"
	"github.com/pkg/errors"
	"github.com/veedubyou/chord-paper-be/go-rewrite/src/errors/api"
	"github.com/veedubyou/chord-paper-be/go-rewrite/src/errors/auth"
	songentity "github.com/veedubyou/chord-paper-be/go-rewrite/src/song/entity"
	songerrors "github.com/veedubyou/chord-paper-be/go-rewrite/src/song/errors"
	songstorage "github.com/veedubyou/chord-paper-be/go-rewrite/src/song/storage"
	userusecase "github.com/veedubyou/chord-paper-be/go-rewrite/src/user/usecase"
	"sync"
)

type Usecase struct {
	db          songstorage.DB
	userUsecase userusecase.Usecase
}

func NewUsecase(db songstorage.DB, userUsecase userusecase.Usecase) Usecase {
	return Usecase{
		db:          db,
		userUsecase: userUsecase,
	}
}

func (u Usecase) GetSong(ctx context.Context, songID string) (songentity.Song, *api.Error) {
	song, err := u.db.GetSong(ctx, songID)
	if err != nil {
		err = errors.Wrap(err, "Failed to get the song from DB")

		switch {
		case markers.Is(err, songstorage.SongNotFoundMark):
			return songentity.Song{}, api.CommitError(err,
				songerrors.SongNotFoundCode,
				"The song can't be found")

		case markers.Is(err, songstorage.SongUnmarshalMark):
			fallthrough
		case markers.Is(err, songstorage.DefaultErrorMark):
			fallthrough
		default:
			return songentity.Song{}, api.CommitError(err,
				api.DefaultErrorCode,
				"Unknown error: Failed to fetch the song")
		}
	}

	return song, nil
}

func (u Usecase) GetSongSummariesForUser(ctx context.Context, authHeader string, ownerID string) ([]songentity.SongSummary, *api.Error) {
	waitgroup := sync.WaitGroup{}

	waitgroup.Add(2)

	var verifyOwnerErr *api.Error
	verifyOwner := func() {
		defer waitgroup.Done()
		verifyOwnerErr = u.userUsecase.VerifyOwner(ctx, authHeader, ownerID)
	}

	var summaries []songentity.SongSummary
	var getSummariesErr error
	getSummaries := func() {
		defer waitgroup.Done()
		summaries, getSummariesErr = u.db.GetSongSummariesForUser(ctx, ownerID)
	}

	go verifyOwner()
	go getSummaries()

	waitgroup.Wait()

	if verifyOwnerErr != nil {
		return nil, api.WrapError(verifyOwnerErr, "Could not verify owner ID")
	}

	if getSummariesErr != nil {
		getSummariesErr = errors.Wrap(getSummariesErr, "Could not fetch song summaries")
		switch {
		case markers.Is(getSummariesErr, songstorage.SongUnmarshalMark):
			fallthrough
		case markers.Is(getSummariesErr, songstorage.DefaultErrorMark):
			fallthrough
		default:
			return nil, api.CommitError(getSummariesErr,
				api.DefaultErrorCode,
				"Ran into issues fetching your songs. Please contact the developer")
		}
	}

	return summaries, nil
}

func (u Usecase) CreateSong(ctx context.Context, authHeader string, song songentity.Song) (songentity.Song, *api.Error) {
	if !song.IsNew() {
		return songentity.Song{}, api.CommitError(
			errors.New("Song has an assigned ID already"),
			songerrors.ExistingSongCode,
			"This song is not in an unsaved state and can't be created")
	}

	if song.Defined.Owner == "" {
		return songentity.Song{}, api.CommitError(
			errors.New("Owner field is empty"),
			auth.WrongOwnerCode,
			"No owner is defined on the song. Check that your login status is valid")
	}

	apiErr := u.userUsecase.VerifyOwner(ctx, authHeader, song.Defined.Owner)
	if apiErr != nil {
		return songentity.Song{}, api.WrapError(apiErr, "Failed to verify song owner")
	}

	song.CreateID()
	song.SetSavedAtToNow()
	createdSong, err := u.db.CreateSong(ctx, song)
	if err != nil {
		return songentity.Song{}, api.CommitError(
			errors.Wrap(err, "Failed to create the song in the DB"),
			api.DefaultErrorCode,
			"Unknown error: Failed to create the song")
	}

	return createdSong, nil
}

func (u Usecase) UpdateSong(ctx context.Context, authHeader string, songID string, song songentity.Song) (songentity.Song, *api.Error) {
	dbSong, apiErr := u.GetSong(ctx, songID)
	if apiErr != nil {
		return songentity.Song{}, api.WrapError(apiErr, "Failed to fetch song")
	}

	freshlyFetchedSong := FreshlyFetchedSong(dbSong)

	apiErr = u.verifySongOwnerBySong(ctx, authHeader, freshlyFetchedSong)
	if apiErr != nil {
		return songentity.Song{}, api.WrapError(apiErr, "Cannot verify that this user owns this song")
	}

	// set these security fields in case someone wants to pull a fast one
	song.Defined.ID = freshlyFetchedSong.Defined.ID
	song.Defined.Owner = freshlyFetchedSong.Defined.Owner

	apiErr = protectSongFromOverwriting(song, freshlyFetchedSong)
	if apiErr != nil {
		return songentity.Song{}, api.WrapError(apiErr, "Song protected from overwriting")
	}

	song.SetSavedAtToNow()

	updatedSong, err := u.db.UpdateSong(ctx, song)
	if err != nil {
		err = errors.Wrap(err, "Failed to update song in DB")
		switch {
		case markers.Is(err, songstorage.SongNotFoundMark):
			return songentity.Song{}, api.CommitError(err,
				songerrors.SongNotFoundCode,
				"The song to be saved doesn't currently exist in the database")
		case markers.Is(err, songstorage.DefaultErrorMark):
			fallthrough
		default:
			return songentity.Song{}, api.CommitError(err,
				api.DefaultErrorCode,
				"Unknown error: Failed to save updated song")
		}
	}

	return updatedSong, nil
}

func protectSongFromOverwriting(songToUpdate songentity.Song, songFromDB FreshlyFetchedSong) *api.Error {
	if songToUpdate.Defined.LastSavedAt == nil {
		err := errors.New("Song payload doesn't have a last saved at field set")
		return api.CommitError(err,
			songerrors.SongOverwriteCode,
			"This song doesn't have a last saved at time, it may not have been created yet. Please upload the initial copy first")
	}

	if songFromDB.Defined.LastSavedAt == nil {
		// unexpected, any song in the DB should have a last saved at
		// but don't block on this
		return nil
	}

	// prevent overwriting a more recent save if the last saved at timestamp is greater than the current one
	// example:
	// A --> B
	//   \----->C
	//
	// Suppose I open the song at time A on two computers, make edits on both somewhat absentmindedly
	// I first save at time B - the payload for the song for the last saved at would be A
	// presume this succeeds, the server copy is now from time B and its last saved at time is also B
	//
	// Now I save another copy at time C, but the predecessor of my copy at time C was from A
	// If I just go with last write wins, then all changes from the copy at time B would be overwritten
	// So by comparing the timestamp at the save at C (A vs B), the save will fail to protect overwriting data.
	//
	// The user can then refetch the copy at time B and copy over their changes from time C (manual merge)
	// and form a copy that will be accepted by the server:
	//
	// A --> B----->B+C
	//   \----->C---/
	if songToUpdate.Defined.LastSavedAt.Before(*songFromDB.Defined.LastSavedAt) {
		err := errors.New("New song has an earlier last saved at than the current last saved time")
		return api.CommitError(err,
			songerrors.SongOverwriteCode,
			"Unable to save - there's been a more recent copy of this song saved and saving it will clobber it. Please try reloading this song in a new tab and copy your work over")
	}

	return nil
}

func (u Usecase) DeleteSong(ctx context.Context, authHeader string, songID string) *api.Error {
	if apiErr := u.verifySongOwnerBySongID(ctx, authHeader, songID); apiErr != nil {
		return apiErr
	}

	err := u.db.DeleteSong(ctx, songID)
	if err != nil {
		err = errors.Wrap(err, "Failed to delete song")
		switch {
		case markers.Is(err, songstorage.SongNotFoundMark):
			return api.CommitError(err,
				songerrors.SongNotFoundCode,
				"The song requested to be deleted couldn't be found")

		case markers.Is(err, songstorage.DefaultErrorMark):
			fallthrough
		default:
			return api.CommitError(err,
				api.DefaultErrorCode,
				"Unknown error: Failed to delete song")
		}
	}

	return nil
}

func (u Usecase) verifySongOwnerBySongID(ctx context.Context, authHeader string, songID string) *api.Error {
	song, apiErr := u.GetSong(ctx, songID)
	if apiErr != nil {
		return api.WrapError(apiErr, "Failed to fetch song")
	}

	return u.verifySongOwnerBySong(ctx, authHeader, FreshlyFetchedSong(song))
}

// this type alias is to make it very explicit that the input is for a song that was just fetched
// from the DB, not a song that was provided by the API caller. since they have the same type
// it can be easy to just thread it through. the caller must acknowledge this difference through an explicit cast
type FreshlyFetchedSong songentity.Song

func (u Usecase) verifySongOwnerBySong(ctx context.Context, authHeader string, song FreshlyFetchedSong) *api.Error {
	apiErr := u.userUsecase.VerifyOwner(ctx, authHeader, song.Defined.Owner)
	if apiErr != nil {
		return api.WrapError(apiErr, "Failed to verify song owner")
	}

	return nil
}
